### PR TITLE
nixos/switch-to-configuration: Document and test handling of socket-activated services

### LINKED ISSUE #73853: Restarting socket units
As mentioned in https://github.com/NixOS/nixpkgs/commit/d37458ad063b04439455fa7c54c3d2abf25a1f5a#diff-0a057d6ff3f6f83f68b859178484f4feR233,
I'd argue the current behavior of restarting sockets is wrong.

# Motivation

One of the features of socket activation [(per Lennart)](http://0pointer.de/blog/projects/socket-activated-containers.html):

> The service's IP socket stays continously connectable, no connection attempt ever fails, and all connects will be processed promptly.

# Existing behavior

If unit changes, socket is restarted along side the unit.

# Expected behavior

If unit changes, socket is not restarted.

cc @abbradar

### PR DESCRIPTION
**Pull request by dasJ**
###### Motivation for this change

More tests, more docs

The manual should explain what this is doing exactly.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests)) → switchTest
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ncfavier** [comment]
@ofborg test switchTest

**Comment by dasJ** [comment]
Also cc @aanderse to remind people that you want the script rewritten ;)

File: nixos/tests/switch-test.nix [hunk 1]
```diff
@@ -675,7 +758,91 @@ import ./make-test-python.nix ({ pkgs, ...} : {
...
+        machine.succeed("[ -S /run/test.sock ]")
+        # Socket-activation of the unit still works
+        if machine.succeed("socat - UNIX-CONNECT:/run/test.sock") != "hello":
+            raise Exception("Socket was not properly activated after the service was restarted")
+
+        # Changing an activated service with stopIfChanged=true and ExecStop= stops the socket
+        # to prevent races. This also changes the reload trigger which is ignored because the
+        # unit needs to be restarted anyway
+        out = switch_to_specialisation("${machine}", "simple-socket-stop-if-changed-and-execstop")
+        assert_contains(out, "stopping the following units: socket-activated.service, socket-activated.socket\n")
```
Comment (roberth) [inline nixos/tests/switch-test.nix:839]: This switches from a non-ExecStop service to an ExecStop one, so it doesn't need to stop the socket yet, but does need to do so on the next switch, even when ExecStop is removed.

File: nixos/doc/manual/development/unit-handling.section.md [hunk 2]
```diff
@@ -48,10 +48,19 @@ checks:
...
-    stopped and started, and the service is stopped and to be started by socket
-    activation.
+    and socket-activated. If `X-StopIfChanged` is **not** set, the service
+    is **restart**ed with the rest.
+
+  - If the socket-activated service does have `X-StopIfChanged` set (the
+    default), the behavior depends on whether `ExecStop=` is set. If set, the
+    `.socket` unit and the `.service` unit are **stop**ped and the `.socket`
+    unit is **start**ed. This is required to prevent races. If `ExecStop=` is
+    not set, there is no need to touch the `.socket` unit. Instead, the
```
Comment (roberth) [inline nixos/doc/manual/development/unit-handling.section.md:62]: This means that despite `stopIfChanged` being enabled, the old service will run until _after_ `etc` has been switched.
I don't think that's acceptable.

**Review by roberth — CHANGES_REQUESTED** [review]
Found a flaw :cry: This will break services where `ExecStop` relies on still having the old `etc`.

I'm not confident that this can be fixed without fixing systemd or mutating the old units. We really need to inhibit the old service start; otherwise it seems we'll have to introduce new corner cases.

**Comment by dasJ** [comment]
@roberth thank you for reviewing the change.
The whole "ExecStop is taken from the new service" is fixable but I won't do that until we agreed what to do with the other points you mentioned.
Running the service in the new environment is imo acceptable because it's only very short and it increases general service availability.
I do however not know each and every service in NixOS and I don't know if services could suffer from this behavior.

Do you have an alternative suggestion I can implement? If we don't come to an agreement I can redo this PR to only clean the code, add test cases, and add docs to at least have the current behavior properly specified and tested.

**Comment by dasJ** [comment]
I removed most of the logic after a longer discussion with Robert. We came to the conclusion that there was no way we can keep up with the race-free nature of the current behavior (please correct me if I got your points wrong).
This would be possible if systemd blessed us with some magic way, but as of writing this comment, it's not.

So this PR mostly adds test cases and proper docs for handling of socket-activated units and leaves the current logic intact.

File: nixos/doc/manual/development/unit-handling.section.md [hunk 3]
```diff
@@ -48,10 +48,11 @@ checks:
...
+    before the activation script is run. This behavior is different when the
+    service is socket-activated, as outlined in the following steps.
 
   - The last thing that is taken into account is whether the unit is a service
-    and socket-activated. Due to a bug, this is currently only done when
-    `X-StopIfChanged` is set. If the unit is socket-activated, the socket is
-    stopped and started, and the service is stopped and to be started by socket
-    activation.
+    and socket-activated. If `X-StopIfChanged` is **not** set, the service
+    is **restart**ed with the rest. If it is set both the service and the
```
Comment (cheriimoya) [inline nixos/doc/manual/development/unit-handling.section.md:56]: "rest" sounds strange here. could be "others" or "other services that are marked to be restarted".

```suggestion
    is **restart**ed with the others. If it is set, both the service and the
```

File: nixos/doc/manual/development/unit-handling.section.md [hunk 4]
```diff
@@ -48,10 +48,11 @@ checks:
     If the flag is enabled, the unit is **stop**ped and then **start**ed. If
```
Comment (cheriimoya) [inline nixos/doc/manual/development/unit-handling.section.md:48]: while you're at it, could you please also change 

```diff
-  - The rest of the behavior is decided whether the unit has `X-StopIfChanged`
-     in the `[Service]` section set (exposed via
+  - Further behavior depends on the unit having `X-StopIfChanged` set in the
+    `[Service]` section (exposed via
```

on line 44-45 (cannot mark it here as it doesn't belong to the change)

File: nixos/modules/system/activation/switch-to-configuration.pl [hunk 5]
```diff
@@ -339,21 +340,23 @@ sub handleModifiedUnit {
                 # If this unit is socket-activated, then stop the
                 # socket unit(s) as well, and restart the
                 # socket(s) instead of the service.
-                my $socketActivated = 0;
+                my $socket_activated = 0;
```
Comment (cheriimoya) [inline nixos/modules/system/activation/switch-to-configuration.pl:343]: why is this snake case when all the other variables are camel case?

File: nixos/tests/switch-test.nix [hunk 6]
```diff
@@ -667,15 +716,79 @@ import ./make-test-python.nix ({ pkgs, ...} : {
...
         assert_contains(out, "would reload the following units: reload-triggers.service, simple-reload-service.service\n")
         assert_contains(out, "would restart the following units: reload-triggers-and-restart-by-as.service, reload-triggers-and-restart.service, simple-restart-service.service, simple-service.service\n")
         assert_lacks(out, "\nwould start the following units:")
-        assert_lacks(out, "as well:")
+
+    with subtest("socket-activated services"):
+        # Socket-activated services don't get started, just the socket
+        machine.fail("[ -S /run/test.sock ]")
+        out = switch_to_specialisation("${machine}", "simple-socket")
+        # assert_lacks(out, "stopping the following units:") nobody cares
```
Comment (cheriimoya) [inline nixos/tests/switch-test.nix:732]: why not throw it out then?

File: nixos/tests/switch-test.nix [hunk 7]
```diff
@@ -699,11 +811,11 @@ import ./make-test-python.nix ({ pkgs, ...} : {
         assert_contains(out, "OnCalendar=2014-03-25 02:59:56 UTC")
         out = switch_to_specialisation("${machine}", "timerModified")
         assert_lacks(out, "stopping the following units:")
+        assert_lacks(out, "NOT restarting the following units:")
         assert_lacks(out, "reloading the following units:")
-        assert_contains(out, "restarting the following units: test-timer.timer\n")
+        assert_contains(out, "\nrestarting the following units: test-timer.timer\n")
```
Comment (cheriimoya) [inline nixos/tests/switch-test.nix:816]: why aren't we stripping away the spaces from each `out` variable (or rather in the `switch_to_specialisation` function)? is it important to check the newline everywhere?

**Review by cheriimoya — CHANGES_REQUESTED** [review]

File: nixos/tests/switch-test.nix (hunk 6 above)
Comment (dasJ) [reply nixos/tests/switch-test.nix:732]: To make it clear that this check was left out explicitly and not just forgotten

File: nixos/tests/switch-test.nix (hunk 7 above)
Comment (dasJ) [reply nixos/tests/switch-test.nix:816]: Because `restarting the following units` and `starting the following units` must somehow be differentiated and using regex would be pretty expensive to run hundreds of times

File: nixos/modules/system/activation/switch-to-configuration.pl (hunk 5 above)
Comment (dasJ) [reply nixos/modules/system/activation/switch-to-configuration.pl:343]: https://metacpan.org/pod/Perl::Critic::Policy::NamingConventions::Capitalization

Comment (cheriimoya) [reply nixos/modules/system/activation/switch-to-configuration.pl:343]: makes sense, but what is `unitsToReload`?

Comment (dasJ) [reply nixos/modules/system/activation/switch-to-configuration.pl:343]: There is currently a mix between camel case (not yet refactored) and snake case (already refactored)

File: nixos/tests/switch-test.nix (hunk 7 above)
Comment (cheriimoya) [reply nixos/tests/switch-test.nix:816]: ah, i didn't know `out` was a string, thought it was an array, my bad

File: nixos/tests/switch-test.nix (hunk 6 above)
Comment (roberth) [inline nixos/tests/switch-test.nix:732]:
```suggestion
        # assert_lacks(out, "stopping the following units:") not relevant
```

**Review by roberth — APPROVED** [review]
Just the suggestions from my comment just now. Lgtm!

**Review by stigtsp — APPROVED** [review]
LGTM
